bson/src/main/org/bson/io/ByteBufferBsonInput.java

```diff
@@ -182,11 +183,64 @@ public void skipCString() {
         buffer.position(pos + length);
     }
 
+    /**
+     * Detects the position of the first NULL (0x00) byte in a 64-bit word using SWAR technique.
+     * <a href="https://en.wikipedia.org/wiki/SWAR">
+     */
     private int computeCStringLength(final int prevPos) {
-        ensureOpen();
-        int pos = buffer.position();
+        int pos = prevPos;
         int limit = buffer.limit();
 
+        // `>>> 3` means dividing without remainder by `Long.BYTES` because `Long.BYTES` is 2^3
+        int chunks = (limit - pos) >>> 3;
+        // `<< 3` means multiplying by `Long.BYTES` because `Long.BYTES` is 2^3
+        int toPos = pos + (chunks << 3);
+        for (; pos < toPos; pos += Long.BYTES) {
+            long chunk = buffer.getLong(pos);
+            /*
+              Subtract 0x0101010101010101L to cause a borrow on 0x00 bytes.
+              if original byte is 00000000, then 00000000 - 00000001 = 11111111 (borrow causes the most significant bit set to 1).
+             */
+            long mask = chunk - 0x0101010101010101L;
+            /*
+              mask will only have the most significant bit set iff it was a 0x00 byte (0x00 becomes 0xFF because of the borrow).
+              ~chunk will have bits that were originally 0 set to 1.
+              mask & ~chunk will have the most significant bit set iff original byte was 0x00.
+             */
+            mask &= ~chunk;
+            /*
+               0x8080808080808080:
+               10000000 10000000 10000000 10000000 10000000 10000000 10000000 10000000
+
+               mask:
+               00000000 00000000 11111111 00000000 00000001 00000001 00000000 00000111
+
+               ANDing mask with 0x8080808080808080 isolates the most significant bit (0x80) in positions where
+               the original byte was 0x00, thereby setting the most significant bit to 1 only at the 0x00 byte position.
+
+               result:
+               00000000 00000000 10000000 00000000 00000000 00000000 00000000 00000000
+                                 ^^^^^^^^
+               The most significant bit is set only at the 0x00 byte position.
```

Comment:
"most significant bit" may refer to the most significant bit in a byte, or in a long (the chunk), and without clarified the text is still unclear. Note also how you used "significant byte **of the long**" below.

```suggestion
               The most significant bit is set in each 0x00 byte, and only there.
```

Reply:
Btw, I am not so sure that "and only there" is true. I suspect, after the first (left-most, closest to the least significant byte) `0x00` byte, bits may be set on other bytes, but we don't care about that, as we are looking only for the left-most one. That's why (again, I suspect) Wikipedia mentions that the technique is applicable only if _"the goal is limited to finding the first zero byte on a little-endian processor"_.

Reply:
Interesting, the claim on Wikipedia about "little-endian only" is misleading (maybe a documentation error or not complete). The algorithm, before `numberOfTrailingZeros`, works on both systems (little or big endian), as the zero-byte detection relies on bitwise operations that are endian-agnostic. The only place where byte endianness matters is in `numberOfTrailingZeros` or `numberOfLeadingZeros` to correctly identify the offset, but Wikipedia does not seem to mention that.

Reply:
> the zero-byte detection relies on bitwise operations that are endian-agnostic

I am not sure that's true, as `chunk - 0x0101010101010101L` is not a bitwise operation, and we actually treat it as if it's a **byte**wise operation. It may not matter, of course, however https://jameshfisher.com/2017/01/24/bitwise-check-for-zero-byte/ mentions:

> But why were we able to treat the subtraction as byte-wise subtraction? The subtraction algorithm doesn’t work like that! Well, it does work like this for our particular case where there are no zero bytes and we are subtracting 1. It is only when doing 00000000 - 00000001 that the carry bit will be set when crossing the byte boundary.

As I mentioned in https://github.com/mongodb/mongo-java-driver/pull/1685#pullrequestreview-2801172065, I don't really understand what's going on (for that, I'd have to spend much more time playing with the code and observing intermediate results, and also learn how subtraction works), but it is not impossible that _"our particular case where there are no zero bytes and we are subtracting 1"_ holds only because we are looking for the left-most zero byte in a long, i.e., there are no other zero bytes before it when the `-` operation is being executed, and so the carry bits (I don't actually know what they are) from them do not exist and can't affect the outcome. The zero byte we need is going to be the left-most in a long only if the byte order is little endian.